Consistent value selector behavior, support double-clicking

Robyt3 opened the pull request:
Make value selector behavior consistent with the generic button logic. Consistently check for completed mouse clicks on the value selector UI element instead of handling some mouse down and mouse up events separately. Fix text mode being activated when moving the mouse over value selectors with held down mouse button. Fix text mode being deactivated immediately when the mouse leaves text area while holding down the mouse button, which is inconvenient when selecting text. Remove unnecessary usage of `CUi::MouseButtonReleased` function.

Support double-clicking value selectors to enter text edit mode in addition to right-clicking. This feels more intuitive to use and also makes it usable without a second mouse button.

Select all text when entering text mode also in the editor, which was previously only a feature in the menus.

Fix value selector edit state being `EDITING` by default instead of `NONE` in the editor and fix maps being marked as modified immediately when a value selector text input is activated. The map will now be marked as modified only when the editing operation is completed, which should be synchonized with the undo history action being added.

Use the `CUi::ConsumeHotkey` function instead of checking the enter keys manually.

Remove unnecessary and unused `CUi::m_ValueSelectorTextMode` variable and its accessors as well as the equivalent `s_TextMode` variable in the editor. This separate flag is unnecessary as we can use the existing `s_pLastTextId` variable instead.

Replace the last remaining usage of the `CUi::MouseButtonReleased` function with `!MouseButton(...)`. The `pMouseSelection->m_Selecting` flag is only set to `true` when the mouse button is already pressed down, so the additional check of `MouseButtonReleased` is unnecessary in this case. In general, this function is an anti-pattern in our UI, as only checking for the mouse button to be released does not guarantee that the mouse was also pressed down over the respective UI element. The `DoButtonLogic` function or similar code should be used instead of only handling the press or release of a mouse button.

This has no effect on #8413.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

archimede67 reviewed (changes requested):
I think this change overall looks good.
However, I encountered two issues:

1. You can trigger the double click to edit even if your first click is outside the input:

https://github.com/ddnet/ddnet/assets/13364635/81246b6a-7f07-40e5-b1e4-1bc608fd9423

I also found out that if you double click but hold the second click to scroll the value, when you release you will enter edit mode

3. You cannot right click another field to enter edit mode if you are editing a field. You first have to exit to then be able to right click to edit another field:

https://github.com/ddnet/ddnet/assets/13364635/9b639049-0b01-45fb-b4d7-0003c8f4b813

Regarding the history capabilities, I believe nothing broke.

Robyt3 commented:
> You can trigger the double click to edit even if your first click is outside the input

That's a general issue with the way we handle double clicks right now. This also affects other UI like double clicking server browser entries. Maybe double click handling should also be part of the UI instead of the engine input so we can handle this properly. We currently rely on SDL entirely to tell us when a click is a double-click. This also doesn't work on Android currently, since the touch events don't tell us the clicks like the mouse events do, so I have to reimplement double-click handling separately anyway for Android support. When double-clicks are handled by the UI we could ensure that both clicks for the double-click were made on the same UI element.

> I also found out that if you double click but hold the second click to scroll the value, when you release you will enter edit mode

Fixed. It will now only enter edit mode with double click if the value was not changed by scrolling.

> You cannot right click another field to enter edit mode if you are editing a field. You first have to exit to then be able to right click to edit another field

Fixed. I think I initially wanted to avoid multiple value selectors being active at the same time for one frame, but this is now fixed differently since the mouse down event stops the current edit mode whereas the mouse up event will start the edit mode on the other UI element.

archimede67 reviewed (approved):
> Maybe double click handling should also be part of the UI instead of the engine input so we can handle this properly.

I agree, but this should be its own PR.
Otherwise, looks good to me!
